instaloader/nodeiterator.py

```diff
@@ -26,7 +27,7 @@
 FrozenNodeIterator.best_before.__doc__ = """Date when parts of the stored nodes might have expired."""
 FrozenNodeIterator.remaining_data.__doc__ = \
     """The already-retrieved, yet-unprocessed ``edges`` and the ``page_info`` at time of freezing."""
-
+FrozenNodeIterator.first_item_timestamp.__doc__ = """Timestamp (UTC) of first item, if an item has been produced."""
```

Comment:
Am I right in that this serves the purpose of `start_of_first_incomplete_run` I proposed in https://github.com/instaloader/instaloader/issues/1206#issuecomment-867472513? Good idea to save it here in the resume file rather than in latest_stamps, since it is a property more-inherent to the resuming mechanism, and this way it will continue to work if multiple resume prefixes are used concurrently. Also seems an elegant way of avoiding the `copy` method with all its drawbacks.

However, could we store the whole first item rather than just its `date_utc` field? Then the NodeIterator class would stay truly generic, which would be handy if we'd ever want a similar mechanism for objects that do not have that timestamp.

Reply:
Yes, that is basically the same as `start_of_first_incomplete_run`.

I remember that I considered saving the whole item, but found some difficulty, but I can't recall exactly what was the problem. Looking again, it seems doable, but instead of `NodeIterator[T]` it would be necessary to limit it to `NodeIterator[JsonExportable]`, since the whole stored object must be stored and loaded as a json string. (Or it could still be any type, but it only gets stored if it happens to be `JsonExportable`.)

Reply:
I think we could store the JSON structure that is returned by Instagram directly, then it is not limited to JsonExportable.

Reply:
The type of object must also be stored, so that when the file contents is read, an object of the correct type can be constructed. And we'd also need a series of `if` clauses to call the proper constructor based on the type. Which is basically what `get_json_structure` and `load_structure` already do. Or do you have another way in mind of saving/restoring the item?

Reply:
We can assume that a FrozenNodeIterator is thawn with a matching NodeIterator, more specifically, one that has been passed the same node_wrapper. Hence, we could assign `self._first_item` the raw `node` object we have in `__next__()`, and let the property `first_item` return `self._node_wrapper(self._first_item)`. I think that should work fine.

Reply:
So, (hopefully) more clearly, we'd have (let's call it `first_node` rather than `first_item`)
- a field `_first_node` in NodeIterator (instead of `_first_item_timestamp`)
- a field `first_node` in FrozenNodeIterator (instead of `first_item_timestamp`), which stores the value of `NodeIterator._first_node`
- assign the raw `node` object to `self._first_node` in `__next__()` at the place where now `self._first_item_timestamp` is assigned
- a property in NodeIterator `first_item()` (I call it first_*item* since what it returns is not a raw node anymore) which returns `self._node_wrapper(self._first_node)` (type `T`)

The callers could then use `iterator.first_item.date_utc` in place of `iterator.first_item_timestamp`.

No need to store the type and still everything stays generic. The node_wrappers are all very simple and efficient constructor functions, which do not do any requests or so.

Reply:
This turned out to be simple and elegant. I've just pushed these changes.